### ISSUE TITLE
CLI sometimes breaks the terminal after game closes

### ISSUE THREAD

**Issue by bitcraft** [Tuxemon/Tuxemon#1071]
**Describe the bug**
On linux, if the CLI interface is enabled, sometimes after the game closes, the terminal will be in a bad state.  The most frequent issue that I have is that no text is visible.  Its still possible to use the terminal.  Can be fixed by using `reset` command, which will reset the terminal and then keystrokes will show normally.

**To Reproduce**
- Open terminal
- Enable CLI in the config file
- Close Tuxemon
- Occasionally, keystrokes will no longer be shown in the terminal

**Expected behavior**
Terminal should not be broken

**Desktop (please complete the following information):**
 - Fedora 35
 - Gnome Terminal
 - Python 3.10

**Comment by bitcraft** [comment]
must have misclicked?  i didn't mean to assign this.  fixed.

**Comment by ultidonki** [comment]
Yeah, just confirming this happens for me too on a different OS. That reset command fixes it, though! Which is much better than closing & reopening a new terminal, like I used to do.

It even breaks when I do nothing, just open and immediately close Tuxemon, when the CLI is enabled. 

**Desktop** 
* Kubuntu 20.04
* Konsole Terminal
* Python 3.8

**Comment by ultidonki** [comment]
The terminal is left in a bad state, if you close pygame using the 'x' in the top right, or the game crashes with an error. 
If you use CTRL+D, the game closes and the terminal works normally.
This makes me think pygame is closing the PromptSession in tuxemon/cli/processor.py, in a messy way that doesn't allow it to cleanup, or something? 
Although if I create a minimal prompt_toolkit example, and then try to cause errors, I can't replicate the terminal misbehaving afterwards. It only seems to happen with Tuxemon. 
 
Also, a specific way to fix the problem is running `stty echo` - this makes the terminal work OK again (echoes the input out again).
Which makes me think prompt_toolkit is turning echo off (using termios?) maybe, and then isn't being cleaned up? But I don't see any issues on their github page where anyone else has the same problem.

I only learned the stuff above while investigating the issue, I'm not an expert on terminals or any of the libraries mentioned, so I'm not sure if the above is true or not, but hopefully it's helpful.

**Comment by bitcraft** [comment]
thanks for your research here.  to complicate matters a bit further, the CLI logic is handled in a different thread, so your thought about prompt_toolkit not cleaning up is probably correct.

**Comment by Divarrek** [comment]
When using a regular thread (not a daemon thread) for the CLI and a `threading.Event()` (or a variable) in the `while` loop condition in processor.py as suggested [in this note](https://docs.python.org/3/library/threading.html#:~:text=Daemon%20threads%20are%20abruptly%20stopped%20at%20shutdown), the terminal doesn't break after closing the game, but the thread stays alive until you send something to the prompt (or exit the python shell console if it's running).

I tried implementing this [here](https://github.com/Tuxemon/Tuxemon/compare/development...Divarrek:Tuxemon:cli?diff=split&w=) but I haven't figured out how to close the PromptSession or the thread completely.

**Comment by JaskRendix** [comment]
@Divarrek I like your approach, even if I'm not sure about removing the `thread.daemon = True`. Regarding the `PromptSession` or the thread, as you said, I would add a stop method inside the processor.py
```
    def stop(self) -> None:
        self.event.set()
```
while in the quit (client.py)
```
    def quit(self) -> None:
        self.exit = True
        self.done = True
        if self.config.cli:
            self.cli.stop()
```
you can also try to remove the break and add `self.event.set()`
```
            except EOFError:
                self.event.set()
            except KeyboardInterrupt:
                print("Got KeyboardInterrupt")
                print("Press CTRL-D to quit.")
                self.event.set()
```


but I need to tell you that I'm wondering if we need to change the "flow", what I mean is moving the event_engine in the `quit` method of the client
```
    if self.config.cli:
        self.cli.stop()
    event_engine = self.session.client.event_engine
    event_engine.execute_action("quit")
```
we could swap out those two calls for just `self.stop()` in the `run()` method. Then, in `stop`, we can add `self.session.close()` to clean up the session
```
    def stop(self) -> None:
        if self.session:
            self.session.close() 
```
To tie it all together, we could add a `close(self)` method inside `session.py` that calls `self.client.quit()`
```
    def close(self) -> None:
        if self.client:
            self.client.quit()
```
That way, everything comes full circle.... What do you think?
